src/Data/Array/Mutable/Unlifted/Linear.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE UnliftedNewtypes #-}
+
+-- |
+-- This module defines a linear wrapper around 'GHC.MutableArray#'
+-- which can be manipulated in pure contexts.
+--
+-- It ensures that effects happen in order. However, access to
+-- out-of-bounds indices causes undefined behaviour.
+--
+-- This module is meant to be imported qualified.
+module Data.Array.Mutable.Unlifted.Linear
+  ( Array#
+  , unArray#
+  , alloc
+  , allocBeside
+  , lseq
+  , size
+  , read
+  , write
+  , copyInto
+  ) where
+
+import Data.Unrestricted.Linear hiding (lseq)
+import Prelude (Int)
+import qualified Prelude as Prelude
+import qualified Unsafe.Linear as Unsafe
+import qualified GHC.Exts as GHC
+
+-- | A mutable array holding @a@s
+newtype Array# a = Array# (GHC.MutableArray# GHC.RealWorld a)
+
+-- | Run an unsafe computation using the underlying MutableArray,
+-- consuming the 'Array#' in process.
+unArray# :: (GHC.MutableArray# GHC.RealWorld a -> b) -> Array# a #-> Ur b
+unArray# f = Unsafe.toLinear (\(Array# a) -> Ur (f a))
+
+-- | Consume an 'Array#'.
+--
+-- Note that we can not implement a 'Consumable' instance because 'Array#'
+-- is unlifted.
+lseq :: Array# a #-> b #-> b
+lseq = Unsafe.toLinear2 (\_ b -> b)
+
+-- | Allocate a mutable array of given size using a default value.
+--
+-- The size should be non-negative.
+alloc :: Int -> a -> (Array# a #-> Ur b) #-> Ur b
+alloc (GHC.I# s) a f =
+  let !new = GHC.runRW# Prelude.$ \st ->
+        case GHC.newArray# s a st of
+          (# _, arr #) -> Array# arr
+   in f new
+
+-- | Allocate a mutable array of given size using a default value,
+-- using another 'Array#' as a uniqueness proof.
+--
+-- The size should be non-negative.
+allocBeside :: Int -> a -> Array# b #-> (# Array# b, Array# a #)
```

Comment:
Ha. You went the extra mile, here. And used unboxed pairs. It's probably wise, fewer lazy things, means fewer opportunities to mix up the effect ordering.

Ah! But you don't really have a choice, really, since `(,)` can only contain lifted values. How convenient.